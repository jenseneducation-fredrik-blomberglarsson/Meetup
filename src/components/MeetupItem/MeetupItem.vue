<template>
  <div class="meetupItem">
    <div class="meetupItemContainer">
      <div class="meetupCard">
        <div class="headerContainer">
          <h3 class="meetupCategory">{{meetup.category}}</h3>
          <h2 class="meetupTitle">{{meetup.title}}</h2>
        </div>
        <div class="content">
          <p class="meetupDesc">{{meetup.desc}}</p>
          <p class="meetupDate">{{meetup.date}}</p>
          <p class="meetupLocation">{{meetup.location}}</p>
          <p class="meetupOrganizer">{{meetup.organizer}}</p>
          <section v-if="!getMeetupToggle" class="meetupListComponent">
            <div
              class="addMeetup"
              v-show="isOn"
              v-on:click="addThisMeetup(meetup); toggle(meetup);"
            >
              <button class="joinMeetupBtn">Attend</button>
            </div>
          </section>
          <section v-if="getMeetupToggle" class="myMeetupListComponent">
            <div class="removeMeetup" v-on:click="removeThisMeetup(meetup);">
              <button class="removeMeetupBtn">Leave</button>
            </div>
          </section>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import { mapActions, mapGetters } from "vuex";
export default {
  name: "MeetupItem",
  props: { meetup: Object },
  // Have to make toggle locally instead of store,
  //because otherwise every item gets toggled on click
  data() {
    return {
      isOn: true,
    };
  },
  computed: {
    ...mapGetters(["getMeetupToggle"]),
  },

  methods: {
    toggle() {
      this.isOn = !this.isOn;
    },
    ...mapActions(["addThisMeetup", "removeThisMeetup"]),
  },
};
</script>

<style scoped>
div {
  display: inline-block;
}
body {
  display: flex;
  justify-content: center;
  align-items: center;
  flex-wrap: wrap;
  flex-direction: row;
}

.meetupItemContainer {
  display: flex;
  justify-content: center;
  align-items: center;
  flex-wrap: wrap;
  padding: 30px;
}

.meetupItemContainer .meetupCard {
  max-width: 300px;
  height: 150px;
  background: #03a9f4;
  margin: 30px 10px;
  padding: 20px 15px;
  display: flex;
  flex-direction: column;
  box-shadow: 0 5px 202px rgba(0, 0, 0, 0.5);
  transition: 0.3s ease-in-out;
}

.meetupItemContainer .meetupCard:hover {
  height: 350px;
  background: linear-gradient(#03a9f4, #f441a5, #ffeb3b, #03a9f4);
  background-size: 400% 400%;
  animation: change 10s ease-in-out infinite;
}

.meetupItemContainer .meetupCard .content {
  margin-top: -140px;
  padding: 10px 15px;
  text-align: center;
  color: #111;
  visibility: hidden;
  opacity: 0;
  transition: 0.3s ease-in-out;
}
.meetupItemContainer .meetupCard:hover .content {
  visibility: visible;
  opacity: 1;
  margin-top: -5px;
  transition-delay: 0.2s;
}
@keyframes change {
  0% {
    background-position: 0 100%;
  }
  50% {
    background-position: 100% 50%;
  }
  100% {
    background-position: 0 100%;
  }
}

.joinMeetupBtn {
  font-family: "Permanent Marker", cursive;
  color: #03a9f4;
  background-color: #d6d2d2;
  width: 100px;
  height: 30px;
  font-size: 1rem;
  text-transform: uppercase;
  text-decoration: none;
  letter-spacing: 2px;
  box-sizing: border-box;
  cursor: pointer;
}

.removeMeetupBtn {
  font-family: "Permanent Marker", cursive;
  color: #03a9f4;
  background-color: #d6d2d2;
  width: 100px;
  height: 30px;
  font-size: 1rem;
  text-transform: uppercase;
  text-decoration: none;
  letter-spacing: 2px;
  box-sizing: border-box;
  cursor: pointer;
}
</style>