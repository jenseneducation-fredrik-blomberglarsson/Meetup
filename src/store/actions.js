import { getMeetups } from "../data/meetupData";

const actions = {
  async fetchMeetups(context) {
    const data = await getMeetups();
    context.commit("setMeetups", data);
  }
}

export default actions;