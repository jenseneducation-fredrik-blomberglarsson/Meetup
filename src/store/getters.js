const getters = {
  getMeetups: state => {
    return state.meetups;
  },
  getJoinedMeetups: state => {
    return state.joinedMeetups;
  },
  getPreviousMeetups: state => {
    return state.previousMeetups;
  },
  getMeetupToggle: state => {
    return state.meetupToggle;
  },
  getBurgerToggle: state => {
    return state.burgerToggle;
  }
}


export default getters;