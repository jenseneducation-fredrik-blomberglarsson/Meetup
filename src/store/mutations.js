const mutations = {
  setMeetups(state, meetups) {
    state.meetups = meetups;
  }
}

export default mutations;