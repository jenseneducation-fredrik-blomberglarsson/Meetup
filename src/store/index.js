import Vue from 'vue'
import Vuex from 'vuex'
import actions from './actions.js';
import mutations from './mutations.js';
import getters from './getters.js';

Vue.use(Vuex)

//Global state
export default new Vuex.Store({
  state: {
    meetups: [],
    joinedMeetups: [],
    previousMeetups: [],
    meetupToggle: false,
  },

  actions: actions,
  mutations: mutations,
  getters: getters
});