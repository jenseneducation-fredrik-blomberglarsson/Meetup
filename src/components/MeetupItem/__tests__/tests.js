import { shallowMount, mount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import store from '../../../store/index.js';
import MeetupItem from '@/components/MeetupItem/MeetupItem.vue';

const localVue = createLocalVue();
localVue.use(Vuex);

describe('The page is rendered', () => {
  // Need to mock MeetupItem here when i've figured out its data
  const mockedMeetupItem = {
    id: '01',
    title: 'Tech Meetup',
    desc: 'Learn front-end development with experienced developers',
    date: 'WED, SEP 23 6:30 PM',
    location: 'StockholmsmÃ¤ssan',
    organizer: 'Stockholm WebNow'
  };

  test('Test so that everything renders properly', async () => {
    const wrapper = shallowMount(MeetupItem, { store, localVue });
    expect(wrapper.element).toMatchSnapshot();
  });

  test('Test to check that output is correct on MeetupItem', async () => {
    const wrapper = shallowMount(MeetupItem, {
      store,
      localVue,
      propsData: {
        mockedMeetupItem
      }
    });

    const meetupTitle = wrapper.find('.meetupTitle');
    const meetupDesc = wrapper.find('.meetupDesc');
    const meetupDate = wrapper.find('.meetupDate');
    const meetupLocation = wrapper.find('.meetupLocation');
    const meetupOrganizer = wrapper.find('.meetupOrganizer');

    expect(meetupTitle.text()).toContain(mockedMeetupItem.title);
    expect(meetupDesc.text()).toContain(mockedMeetupItem.desc);
    expect(meetupDate.text()).toContain(mockedMeetupItem.date);
    expect(meetupLocation.text()).toContain(mockedMeetupItem.location);
    expect(meetupOrganizer.text()).toContain(mockedMeetupItem.organizer);
  });

  test('Check if joinMeetup is displayed if joinMeetupToggle is false', async () => {
    const wrapper = shallowMount(MeetupItem, {
      store,
      localVue,
      propsData: {
        mockedMeetupItem
      }
    });

    const button = wrapper.find('.joinMeetupBtn');

    expect(store.state.joinMeetupToggle).toBe(false);
    expect(button.exists()).toBe(true);
  })

  test('Check if leaveMeetup is displayed if joinMeetupToggle is true', async () => {
    const wrapper = shallowMount(MeetupItem, {
      store,
      localVue,
      propsData: {
        mockedMeetupItem
      }
    });
    store.state.joinMeetupToggle = true;
    await wrapper.vm.$nextTick();
    const button = wrapper.find('.joinMeetupBtn');

    expect(store.state.joinMeetupToggle).toBe(true);
    expect(button.exists()).toBe(true);
  })

  test('Check that joinMeetup is called on button click', async () => {
    const state = {
      isMeetupToggleActive: false
    }

    const actions = {
      joinMeetup: jest.fn()
    };
    const getters = {
      getMeetupToggle: state => {
        return state.isMeetupToggleActive;
      }
    };
    const store = new Vuex.Store({
      actions,
      getters,
      state
    });

    const wrapper = shallowMount(MeetupItem, {
      store,
      localVue,
      propsData: {
        mockedMeetupItem
      }
    });

    const button = wrapper.find('.joinMeetupBtn');
    await button.trigger('click');
    await localVue.nextTick();
    expect(actions.joinMeetup).toHaveBeenCalled();
  })

  test('Check that leaveMeetup is called on button click', async () => {
    const state = {
      isMeetupToggleActive: true
    }

    const actions = {
      leaveMeetup: jest.fn()
    };
    const getters = {
      getMeetupToggle: state => {
        return state.isMeetupToggleActive;
      }
    };
    const store = new Vuex.Store({
      actions,
      getters,
      state
    });

    const wrapper = shallowMount(MeetupItem, {
      store,
      localVue,
      propsData: {
        mockedMeetupItem
      }
    });

    const button = wrapper.find('.joinMeetupBtn');
    await button.trigger('click');
    await localVue.nextTick();
    expect(actions.leaveMeetup).toHaveBeenCalled();
  })
})