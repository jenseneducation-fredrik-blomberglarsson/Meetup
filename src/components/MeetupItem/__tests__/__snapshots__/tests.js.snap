// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The page is rendered Test so that everything renders properly 1`] = `
<div
  class="meetupItem"
>
  <div
    class="meetupItemContainer"
  >
    <div
      class="meetupCard"
    >
      <div
        class="headerContainer"
      >
        <h3
          class="meetupCategory"
        >
          Tech
        </h3>
         
        <h2
          class="meetupTitle"
        >
          Learn front-end from scratch
        </h2>
      </div>
       
      <div
        class="content"
      >
        <p
          class="meetupDesc"
        >
          Learn front-end development with experienced developers
        </p>
         
        <p
          class="meetupDate"
        >
          WED, SEP 23 6:30 PM
        </p>
         
        <p
          class="meetupLocation"
        >
          Online
        </p>
         
        <p
          class="meetupOrganizer"
        >
          Stockholm WebNow
        </p>
         
        <section
          class="meetupListComponent"
        >
          <div
            class="addMeetup"
          >
            <button
              class="joinMeetupBtn"
            >
              Attend
            </button>
          </div>
        </section>
         
        <!---->
      </div>
    </div>
  </div>
</div>
`;
